test/e2e/smoke_test.go

```diff
@@ -0,0 +1,141 @@
+//+build e2e
+
+package e2e
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strconv"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	templatev1 "github.com/openshift/api/template/v1"
+	routev1client "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+	templatev1client "github.com/openshift/client-go/template/clientset/versioned/typed/template/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const GOOD_TEMPLATE = "nodejs-mongodb-example"
+const BAD_TEMPLATE = "non-existent"
+const REGEX = `\<span.*id=\"count-value\"\>(?P<counter>\d+)\<\/span\>`
+
+var _ = Describe("Openshift on Azure end user e2e tests [SmokeTest]", func() {
+	defer GinkgoRecover()
+
+	Context("When the template exists", func() {
+		It("should deploy a template and check the visit counter increments", func() {
+			// Create the template
+			templateclient, err := templatev1client.NewForConfig(c.config)
+			Expect(err).NotTo(HaveOccurred())
+			template, err := templateclient.Templates("openshift").Get(
+				GOOD_TEMPLATE, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			// Create a secret to hold the template params
+			coreclient, err := corev1client.NewForConfig(c.config)
+			Expect(err).NotTo(HaveOccurred())
+			secret, err := coreclient.Secrets(c.namespace).Create(&corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "parameters",
+				},
+				StringData: map[string]string{
+					"NAME":         GOOD_TEMPLATE,
+					"MEMORY_LIMIT": "1024Mi",
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			// Instantiate the template
+			ti, err := templateclient.TemplateInstances(c.namespace).Create(
+				&templatev1.TemplateInstance{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "templateinstance",
+					},
+					Spec: templatev1.TemplateInstanceSpec{
+						Template: *template,
+						Secret: &corev1.LocalObjectReference{
+							Name: secret.Name,
+						},
+					},
+				})
+			Expect(err).NotTo(HaveOccurred())
+
+			// Wait for the deployments to finish
+			watcher, err := templateclient.TemplateInstances(c.namespace).Watch(
```

Comment:
Instead of a watch I would vote for polling since it is less code here and simpler to maintain. Add a `func (t *testclient) templateInstanceIsReady (bool, error)` function in conditions_test.go and wait for it here with a `wait.PollImmediate(2*time.Second, time.Minute, t.templateInstanceIsReady)`

Reply:
This comment got buried in the changes I pushed: @kargakis I'll ping you tomorrow to go over this, I'm a bit confused on how exactly to do that.